src/cargo/ops/cargo_add/mod.rs

```diff
@@ -329,6 +338,26 @@ pub struct DepOp {
     pub tag: Option<String>,
 }
 
+fn spec_fix_suggestion(arg: &DepOp) -> Option<&'static str> {
+    let spec = arg.crate_spec.as_deref()?;
+
+    if arg.git.is_none()
+        && (spec.starts_with("git@") || spec.starts_with("ssh:") || spec.ends_with(".git"))
+    {
+        Some("git URLs must be specified with --git")
+    } else if arg.registry.is_none()
+        && (spec.starts_with("registry+") || spec.starts_with("sparse+"))
+    {
+        Some("registy can be specified with --registry")
+    } else if spec.contains("://") {
+        Some("`cargo add` expects crates specified as 'name' or 'name@version', not URLs")
```

Comment:
Can't this be a git url as well?

Reply:
Yes, that's why the text is phrased to fit use along `--git` too (e.g. in case they swapped url with crate names)